Borders: Switch to ToolsPanel for displaying UI

aaronrobertshaw opened the pull request:
Depends on:
- ~https://github.com/WordPress/gutenberg/pull/32392~

Related: 
- #36641
- https://github.com/WordPress/gutenberg/issues/27331
- https://github.com/WordPress/gutenberg/pull/33744
- https://github.com/WordPress/gutenberg/pull/34027
- https://github.com/WordPress/gutenberg/pull/34061
- https://github.com/WordPress/gutenberg/pull/36753

## Description

This PR:
- Adds new `border` InspectorControls group slot
- Updates the border block support hooks to provide callbacks required for `ToolsPanel`
- Updates the border hook to inject border support controls via the new InspectorControls slot
- Updates the global styles border panel to use the new `ToolsPanel`
- Adds default CSS styles to ensure visual borders when width, style or color are set individually

### Default Controls

_This PR only addresses the switch to the new `ToolsPanel` component. Deciding which controls should display by default for which blocks will be addressed via a separate PR._

## How has this been tested?

Manually.

#### Test instructions

1. Edit a post, add a group block, then select that block.
2. Confirm that the border controls within the sidebar are within a ToolsPanel
3. Confirm each border control works appropriately (Note: this PR does not make any specific block's border controls "default" in terms of always displaying within the new panel)
4. Switch to the site editor
5. Under "By block type", expand the group block panel and confirm the border controls are also in the new panel there
6. Test the border controls within global styles and ensure they work as expected.

## Screenshots <!-- if applicable -->

https://user-images.githubusercontent.com/60436221/129139274-4dc35162-1e91-4c20-bd14-9a6cf5a91ae9.mp4

## Types of changes
New feature. Enhancement.

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/javascript/ -->
- [x] My code follows the accessibility standards. <!-- Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/accessibility/ -->
- [x] I've tested my changes with keyboard and screen readers. <!-- Instructions: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/accessibility-testing.md -->
- [x] My code has proper inline documentation. <!-- Guidelines: https://developer.wordpress.org/coding-standards/inline-documentation-standards/javascript/ -->
- [x] I've included developer documentation if appropriate. <!-- Handbook: https://developer.wordpress.org/block-editor/ -->
- [x] I've updated all React Native files affected by any refactorings/renamings in this PR (please manually search all `*.native.js` files for terms that need renaming or removal). <!-- React Native mobile Gutenberg guidelines: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/code/native-mobile.md -->

aaronrobertshaw committed:
Update hooks with new utils for progressive display

aaronrobertshaw committed:
Update border support to use ToolsPanel

Update border support to display via progressive disclosure panel

Updating border support UI styling for grid layout

Update border global styles panel

Fix border global style hasValue checks

ramonjd commented:
I just took this for a test run. It works well for border supports in both the block and site editors 🎉 

<img width="1114" alt="Screen Shot 2021-08-14 at 7 15 31 am" src="https://user-images.githubusercontent.com/6458278/129419214-cf33bbbf-313e-457c-b1a5-1f71f52adcad.png">


I know you're coming back to this after all the changes to the ToolsPanels, but as far as I could see there wasn't much more than updating the comments to refer to `ToolsPanel` instead of progressive disclosure and using `cleanEmptyObject` for the reset functions ??

There' a WIP defaults PR for borders over at https://github.com/WordPress/gutenberg/pull/34061

ramonjd commented on packages/block-editor/src/hooks/border-color.js:
```diff
@@ -81,6 +85,39 @@ export function BorderColorEdit( props ) {
+
+	setAttributes( {
+		borderColor: undefined,
+		style: removeBorderAttribute( style, 'radius' ),
```
Do we need this line here? 

Resetting `borderColor` is enough?

aaronrobertshaw replied:
It isn't, the line below contains a copy/paste error. 

✅ Fixed.

ramonjd commented on packages/block-editor/src/hooks/border.js:
```diff
@@ -42,22 +62,73 @@ export function BorderPanel( props ) {
+			border: undefined,
+		} );
+
+		props.setAttributes( { style: newStyle } );
```
```suggestion
		props.setAttributes( { style: newStyle, borderColor: undefined } );
```

aaronrobertshaw replied:
I'd checked that `borderColor` was removed from the attributes in the code view of the editor. It is, despite missing the attribute being reset here. I believe that is a result of adjusting border colors when the width changes. 

Thanks for catching this one!

aaronrobertshaw committed:
Fix typo in border color reset

aaronrobertshaw committed:
Incorporate color control styles for ToolsPanel and remove border styles

aaronrobertshaw commented:
Thanks for reviewing this @ramonjd. Appreciate it 👍 

> I know you're coming back to this after all the changes to the ToolsPanels, but as far as I could see there wasn't much more than updating the comments to refer to ToolsPanel instead of progressive disclosure 

Looks like I'd missed updating the border panel's CSS that referred to the `.components-progressive-disclosure-panel` during the last rebase. With tweaks to the `ToolsPanel` styles the `border.scss` file can be removed now. The new styles overlap changes that were required for the color block supports panel so there'll be less in the long run.

In terms of the comments, I see references to a progressive discovery panel which is what the ToolsPanel is. The possibility of renaming the ToolsPanel has come up elsewhere so for the time being, maybe we should leave these comments. The nature of the panel won't change but the component's name might.

> and using cleanEmptyObject for the reset functions ??

Thanks for catching the copy/paste error in the `border-color` hook. 

It should be `style: removeBorderAttribute( style, 'color' )`. There is a call to clean the style object in the `removeBorderAttribute` util function so that's covered.

aaronrobertshaw committed:
Add missing reset of borderColor attribute to resetAll

ramonjd reviewed (approved):
This is working well for me with the defaults in https://github.com/WordPress/gutenberg/pull/34061 activated.

Independent of any blocking concerns/discussions, of which I'm not aware, this LGTM. 

Thank you!

✅ I can apply and reset border support controls as expected. The corresponding inline styles are applied/removed.
✅ All changes are reflected on the frontend
✅ Updates to border styles in the Site Editor are applied to selected blocks

![fern-border-support](https://user-images.githubusercontent.com/6458278/129821830-fa1f0380-01ec-4245-a537-9bc4dcd52bbe.gif)

aaronrobertshaw committed:
Move creation of ToolsPanel to within fills

Creating the ToolsPanel for block supports around the slot resulted in the slot adding an extra div within the ToolsPanel breaking its layout. This div is needed for the slot's ref to be applied allowing virtual bubbling.  This change effectively moves the ToolsPanel inside that. It also avoids needing to retrieve ToolsPanelContext and recreating the context provider.

aaronrobertshaw committed:
Revert "Move creation of ToolsPanel to within fills"

This reverts commit 94f444c15c809d1932d04adc35d7d0e130bc3dbb.

aaronrobertshaw committed:
Remove unused header prop

aaronrobertshaw committed:
Fix layout when slot injects wrapping div inside tools panel

aaronrobertshaw committed:
Update border support use of ToolsPanel

These changes include adding a new border group to the inspector controls slots and updating the border tools panel to match the latest changes to that component.

aaronrobertshaw commented:
This PR has been rebased onto https://github.com/WordPress/gutenberg/pull/34725 which updates the new InspectorControls slots for block supports to use virtual bubbling. This PR has also been updated so the border support hook injects controls via the `border` slot instead of creating the `ToolsPanel` itself.

stacimc committed:
Linting

ciampo commented on packages/block-editor/src/hooks/border.js:
```diff
@@ -39,22 +57,80 @@ export function BorderPanel( props ) {
+		'__experimentalDefaultControls',
+	] );
+
+	const createResetAllFilter = (
+		borderAttribute,
+		topLevelAttributes = {}
+	) => ( newAttributes ) => ( {
+		...newAttributes,
+		...topLevelAttributes,
+		style: {
+			...newAttributes.style,
+			border: {
+				...newAttributes.style?.border,
+				[ borderAttribute ]: undefined,
+			},
+		},
+	} );
```
Just a thought — should the reset all logic be applying/composing the individual `reset*` functions instead? (e.g. `resetBorderColor`, `resetBorderRadius`, `resetBorderStyle`, `resetBorderWidth`)

I see how it could be slightly less performant, but the new approach would avoid the duplication of the "resetting" logic in this file — WDYT?

aaronrobertshaw replied:
I'm not sure I understand what you are saying sorry. Are you asking whether we should be using the `resetAll` callback on the `ToolsPanel` itself here?

If so, we can't do that because these `ToolsPanelItem`s are being added to a `ToolsPanel` via a SlotFill. The slots provided by the grouped `InspectorControls` get wrapped in a `ToolsPanel`. That panel won't know what controls have been added inside the panel so each item needs to passing long what it needs to change in the attributes object when reset.

ciampo replied:
Got it, I had overlooked the fact that the `ToolsPanelItem`s are being injected via SlotFill 😅 

I have another question then — could the `createResetAllFilter` function take a callback as an argument, instead of the `borderAttribute`? Something like `createResetAllFilter( resetBorderColor )` instead of `resetAllFilter={ createResetAllFilter( 'color', { borderColor: undefined, } ) }` ?

Excuses in advance in case this makes little sense.

aaronrobertshaw replied:
Maybe I'm being blind at the moment but I'm not seeing how the callback suggestion cleans up the current approach. 

We can't use `resetBorderColor` as that actually updates block attributes before all the other filters have been applied. In terms of a new callback, we'd be passing a callback to a higher-order function that's purpose is to create a callback. 

We could simplify `createResetAllFilter`by not having to deal with the top-level block attribute the border color needs to reset. It just means we need a custom function for the border color defined which makes things a touch more verbose.

<details>
  <summary>Here's a diff for what I mean</summary>
  
```diff
diff --git a/packages/block-editor/src/hooks/border.js b/packages/block-editor/src/hooks/border.js
index 83f0892b5c..c37dbeae03 100644
--- a/packages/block-editor/src/hooks/border.js
+++ b/packages/block-editor/src/hooks/border.js
@@ -62,12 +62,8 @@ export function BorderPanel( props ) {
 		'__experimentalDefaultControls',
 	] );
 
-	const createResetAllFilter = (
-		borderAttribute,
-		topLevelAttributes = {}
-	) => ( newAttributes ) => ( {
+	const createResetAllFilter = ( borderAttribute ) => ( newAttributes ) => ( {
 		...newAttributes,
-		...topLevelAttributes,
 		style: {
 			...newAttributes.style,
 			border: {
@@ -111,8 +107,16 @@ export function BorderPanel( props ) {
 					label={ __( 'Color' ) }
 					onDeselect={ () => resetBorderColor( props ) }
 					isShownByDefault={ defaultBorderControls?.color }
-					resetAllFilter={ createResetAllFilter( 'color', {
+					resetAllFilter={ ( newAttributes ) => ( {
+						...newAttributes,
 						borderColor: undefined,
+						style: {
+							...newAttributes.style,
+							border: {
+								...newAttributes.style?.border,
+								color: undefined,
+							},
+						},
 					} ) }
 					panelId={ clientId }
 				>
```
</details>


I've looked at this so many times I'm probably not seeing the forest for the trees. I don't feel strongly about the approach, so I'm happy to change this to whatever is needed to move this forward.

ciampo replied:
Thank you for the explanation once again. I think the confusion here mostly come from me not fully grasping how the `resetAllFilter` is supposed to work.

> I don't feel strongly about the approach, so I'm happy to change this to whatever is needed to move this forward.

We can leave this as-is, and sorry if my review slowed this PR down more than it should have.

ciampo commented on packages/block-editor/src/hooks/border.js:
```diff
@@ -39,22 +57,80 @@ export function BorderPanel( props ) {
+		<InspectorControls __experimentalGroup="border">
+			{ isWidthSupported && (
+				<ToolsPanelItem
+					className="single-column"
```
I'm still catching up with the changes happened in the past 2 weeks, but I believe that we're moving away from the `single-column` className?

aaronrobertshaw replied:
We did in the `ToolsPanel` itself. 

However, we still need a means of managing which controls span what columns. I did have a PR switching the `ToolsPanel` to expose grid or grid item properties to manage the layout. It was [requested](https://github.com/WordPress/gutenberg/pull/35621#issuecomment-946071523)  to avoid that for now until more discussion had taken place around just exactly what props we wished to expose.

I believe @stacimc has split the updating of border panel layout into a new PR ( https://github.com/WordPress/gutenberg/pull/35938 ) rather than updating this one directly. This `single-column` class name is a remnant from the old approach in the `ToolsPanel` however it is used to make those controls only span a single column in this panel [here](https://github.com/WordPress/gutenberg/pull/35938/files#diff-32de6bc912076f3b778ba2b7a6cafcc835596c521715d0bc644618fd8777b80eR1-R5).

ciampo replied:
This makes sense! I think I got confused by the fact that the className is the same — but it looks like the styles now come from _outside_ the `ToolsPanel` component. All good!

ciampo commented on packages/block-editor/src/components/inspector-controls/slot.js:
```diff
@@ -34,6 +34,7 @@ export default function InspectorControlsSlot( {
 			<BlockSupportToolsPanel group={ group } label={ label }>
 				<BlockSupportSlotContainer
 					{ ...props }
+					className="components-tools-panel__items-wrapper"
```
Why is this className necessary?

aaronrobertshaw replied:
Good catch, thank you!

I don't believe it is needed anymore, I missed this while repeatedly rebasing this chain of PRs. Thankfully it's not 3 or 4 deep now. 😅 

The class was removed in https://github.com/WordPress/gutenberg/pull/34725/commits/e5809a887ef94fa9b90f896d6f3436fa3e325da7 as part of the work making the grouped `InspectorControls` slots bubble virtually for block supports.

ciampo replied:
Cool! Then I guess it can be safely removed from this PR, right?

ciampo reviewed (commented):
Not sure about the status of this PR (I've been AFK for a couple of weeks), but I thought I'd give it another look since it's still open :) Feel free to ignore my feedback in case it's not relevant!

aaronrobertshaw committed:
Remove unneeded classname from slot container

ciampo reviewed (approved):
Thank you @aaronrobertshaw  for your patience in answering my review feedback. Let's ship this! 🚀

stacimc committed:
Border panel: Update panel layout (#35938)

* Restore single-column styling for appropriate controls

* Remove clear button from Color picker in ToolsPanel

stacimc commented:
Just merged the layout changes from #35938 into this branch for simplicity 👍

aaronrobertshaw commented:
> Just merged the layout changes from #35938 into this branch for simplicity 👍

Thanks @stacimc 👍 

I'll get this updated as necessary to ensure the border-color control displays the correct palettes. Then we can get a final review/approval.

aaronrobertshaw commented:
This has been updated to display multiple origin color palettes for the border color control. 

<img width="280" alt="Screen Shot 2021-11-17 at 4 08 42 pm" src="https://user-images.githubusercontent.com/60436221/142144670-5dea4cb1-f442-4f59-83c0-3181f0a8a19d.png">


It does this through [extracting the retrieval of colors and gradients settings](https://github.com/WordPress/gutenberg/pull/34027/files?authenticity_token=sp5GSHUic9vbzW8R9mz0rnTd9gElUGu0cII5wvQunLhbJtyrnGZpXdnzVde7Cw0BfepgC1JlN76UpGZiNQ6D8A%3D%3D&file-filters%5B%5D=.js#diff-85a751ab9d4e3ebfa54b7e8301039f3704c0808eefb45179257328f08025f33c) into a custom hook as per the PR [switching the color support switch to the `ToolsPanel`](https://github.com/WordPress/gutenberg/pull/34027). I tested this with a few different classic and block-based themes and didn't run into any problems.

Let me know if you see any issues with the approach taken.

ciampo reviewed (approved):
Code changes LGTM 🚀 

although I didn't have time to test this thoroughly in the editor. It'd be probably best if someone could test the changes before merging. Thank you @aaronrobertshaw and @stacimc for working on this!

ramonjd commented:
Retested this just now. It's working as expected for me 🎉 Thank you!

> (Note: this PR does not make any specific block's border controls "default" in terms of always displaying within the new panel)

In https://github.com/WordPress/gutenberg/pull/34061, where we set the defaults, maybe we should show both width and style by default at least wherever we're opting into width until https://github.com/WordPress/gutenberg/pull/34465 or https://github.com/WordPress/gutenberg/pull/34178 get us further.

I was absent-mindedly caught out again when adjusting the border width, expecting to see a border. 

![Nov-18-2021 08-03-24](https://user-images.githubusercontent.com/6458278/142282144-55d9aa15-7945-4698-91ac-33699db1d0bf.gif)

aaronrobertshaw commented:
> In #34061, where we set the defaults, maybe we should show both width and style by default at least wherever we're opting into width until #34465 or #34178 get us further.

That sounds like a reasonable compromise to keep things moving and reduce work in keeping this and #34061 up-to-date. I'd say the ideal case is that we can land #34465 / #34178 first but they could be a while away from merging.

As it is now in the old panel, all the border controls are shown by default. So perhaps adding width and style as defaults in the `ToolsPanel` isn't much of a shift at all and is ok.

@jasmussen or @mtias do you have any thoughts on the best path forward on switching the border support panel to the `ToolsPanel` as discussed here?

jasmussen commented:
The biggest challenge with border controls are those that Ramon notes: when I adjust the border width, I expect to see a change reflected visually.

<img width="492" alt="Screenshot 2021-11-18 at 09 02 13" src="https://user-images.githubusercontent.com/1204802/142375808-a268633c-7ec7-4d0e-9e52-0d32ed422f56.png">

But with how it works, of course, I need to set at least a style as well. That's not a great experience, and I would agree it's made more obtuse by the toolspanel, and needing to add both border width and border style.

It sounds like #34465 is an attempt at addressing that, but it also seems like it still considers border-width and border-style to be separate properties. Can we treat them as one unit? When you add a "Border" control, you get both width and style at the same time? Although they are separate properties in CSS, that doesn't need to be the case in the UI control. Along the same lines, adjusting a border width should also set the solid style if it isn't set already. 

Color might still be a control you add separately, especially since you can apply a `border: 1px solid;` and it inherits the `currentColor` of the text, which by default feels like a good enough experience. 

The following 3 minute doodle looks a bit to the future and assumes we keep a dedicated "Border" panel:
<img width="526" alt="Screenshot 2021-11-18 at 09 16 10" src="https://user-images.githubusercontent.com/1204802/142377736-35fe81c5-7b88-4bfe-9f3f-9ae83f617516.png">

However the only thing I like about that one is that it embraces the ItemGroup component (#35093) which we should probably be using for every color property, so we don't have to show the entire core/theme palette and color card for every property out there. But speaking of mockups, it would be good to check in with both @shaunandrews, @javierarce and @jameskoster as I believe they've all looked into how borders could work.

---

What's next?

At the moment, hiding border controls one by one in the tools panel doesn't feel like an upgrade from the collapsed-by-default border panel:
<img width="456" alt="Screenshot 2021-11-18 at 09 02 52" src="https://user-images.githubusercontent.com/1204802/142378224-f6f3ffc5-0aed-405b-b68a-d46567cb88a7.png">

In that vein, it might be worth waiting to see what #34465 can bring to the table. 

Alternately, can Border width and style be grouped into a single item that outputs a `border: [border-width]px [style];` without the color? That might be a path forward otherwise.

jameskoster commented:
I feel as though width, style, and color should be mandatory controls for this panel. I can't think of an occasion where I'd want a border without a width or a style, and very (very) few where I wouldn't also want a color.

Edit: Here's how I'd expect this panel to work:


https://user-images.githubusercontent.com/846565/142402159-4616a8cd-62e7-4d6c-a9b5-a7c8da435d3e.mp4

stacimc commented:
> Along the same lines, adjusting a border width should also set the solid style if it isn't set already.

This was my goal in #34465: when width or color is adjusted, the solid style is automatically selected if there's no existing style selection. The problem is that we need to make sure to not override selections from global styles or `theme.json` in the process, and for that we need #34178, which adds an API to access global styles within the blocks, but appears to be stalled.

I do have another approach open in #34131, which uses the `:where` selector to add a base `border: 0 solid currentColor` style with 0 priority. Maybe that could get us closer to what we want in the meantime?

> Alternately, can Border width and style be grouped into a single item that outputs a border: [border-width]px [style]; without the color?

I'm having trouble imagining what this would look like -- this would be a single control in the UI that combines width and style?

ramonjd commented:
>  I can't think of an occasion where I'd want a border without a width or a style, and very (very) few where I wouldn't also want a color.

👍 We can try this out when testing the default over at https://github.com/WordPress/gutenberg/pull/34061

I think I agree: even in CSS, where there is no border it's awfully convenient to do something like `border: 1px solid pink`, that is, defining all three properties in one go.

stacimc commented:
Rebased to bring in latest changes to the multiple origin color palettes.

The lack of multiple palettes for border color is listed as a must-have for 5.9 in #36641, so I also cherry-picked @aaronrobertshaw's fixes for that issue into a separate PR without the ToolsPanel changes, in case that is necessary: #36753

aaronrobertshaw committed:
Try adding some default fallbacks

When a user first sets a border width or color their may not be a border style set. These changes aim to provide a fallback for that (in the frontend/block editor) or apply a border style by default in the site editor for Global Styles.

aaronrobertshaw commented:
In https://github.com/WordPress/gutenberg/pull/33743/commits/8956b4969d6724c343093c339cabb759fa27b4a3 I've trialled a new approach to minimise the issue where a user could select a border width or color and see no visual border.

For the block editor and individual blocks, there are a couple of new styles with zero specificity, thanks to `:where`, which apply a `solid` border style when required. This allows this default style to take effect only when no Global Styles have been configured whether by the user or theme.

In the site editor, the same approach as the block editor and frontend won't work. However, given this is where Global Styles are managed and it is aware of theme supplied values, we can safely apply a default border style directly within the Global Styles border panel when selecting a color or width.

The idea is that this approach is sufficient to unblock the switch to the ToolsPanel for the border block support panel. Let me know if there are any great issues with this approach that would need to be addressed first.

#### Block Edtior 

https://user-images.githubusercontent.com/60436221/143190830-bf31e207-914c-476a-b54d-f125ddfa3d37.mp4

#### Site Edtior

https://user-images.githubusercontent.com/60436221/143190908-3eb366d1-749b-4504-9b5a-7c7aa8553d29.mp4

aaronrobertshaw commented:
> > Alternately, can Border width and style be grouped into a single item that outputs a border: [border-width]px [style]; without the color?
> 
> I'm having trouble imagining what this would look like -- this would be a single control in the UI that combines width and style?

I'm also not sure what's really being proposed here in terms of the control itself. 

- How will resetting the individual or combined values work? 
- What if the user only wants to clear the width? 
- How does the control set the initial border-style value until we have access to the Global Styles values via [#34178](https://github.com/WordPress/gutenberg/pull/34178)?

> I do have another approach open in #34131, which uses the :where selector to add a base border: 0 solid currentColor style with 0 priority. Maybe that could get us closer to what we want in the meantime?

#34131 is similar to the approach I'd tried a while ago in [#31366](https://github.com/WordPress/gutenberg/pull/31366/files#diff-b9209e44c092b9376aa2b9afbbb9c4cc64429d2870f99334e478f8f2e653a336R100-R108) to adopt the border support for the image block. 

Here, I've made it such that we only tweak the border style when a border color or width selection has been made. Using `border: 0 solid currentColor` means that selecting a border color alone still won't show a visual border as the width when not set explicitly is `0` rather than picking up the browser default. I also found that it could be possible for a block to not have a CSS class matching `*="wp-block"`.

The current approach, as per https://github.com/WordPress/gutenberg/pull/33743/commits/8956b4969d6724c343093c339cabb759fa27b4a3, should mean that we can use the ToolsPanel without having to force border controls on as defaults if we don't wish to. If any individual control out of width, color, or style is toggled on for display and set, a border will show.


https://user-images.githubusercontent.com/60436221/143196037-d273436b-9078-401c-9db2-ec100871c946.mp4

jasmussen commented:
Thanks for your patience, as usual, and you ask excellent questions. 

To take stock a bit: the tools panel can benefit border in that it does add explicit ways to clear/unset a border, and with the change to show a plus in an empty tools panel, it provides a good flow for intentionally _adding_ a border.

The shortcoming with borders not showing up unless at least style is set was the key problem to solve. Impressively it seems like you [made some progress on that](https://github.com/WordPress/gutenberg/pull/33743#issuecomment-977593105)! How do you feel about that solution, and what does it mean for specificity in case, say, a theme had defined preset borders using theme.json?

[Jay's feedback is still valid](https://github.com/WordPress/gutenberg/pull/33743#issuecomment-972743640), though: if you're adding a border, you'll most likely also want to customize the color at least. That's what makes me feel like this is a single control, not multiple separate ones. Which circles us back to your question:

> How will resetting the individual or combined values work?

I don't think we need to optimize for clearing individual values. You can clear the entire border by unchecking it from the ellipsis, or if need be, backspace out a value. But ultimately, just like how adding a border should be a single step that includes width, style, and color, so could clearing it again be one step.

aaronrobertshaw commented:
> The shortcoming with borders not showing up unless at least style is set was the key problem to solve. Impressively it seems like you made some progress on that! How do you feel about that solution, and what does it mean for specificity in case, say, a theme had defined preset borders using theme.json?

I'm not sure the solution is 100% perfect but I believe it is definitely solid enough to proceed with and then iterate on.

For the block editor and frontend, I think the solution is ok. The new styles have zero specificity so any user set global styles or styles coming from theme.json will override them. I also don't see theme authors setting `border-style: none` via theme.json, they would likely in such cases, just disable the border UI.

As for the Global Styles, the site editor already has access to the theme.json provided styles. So there, the border panel will only set a "default" border style of `solid` if the theme.json hasn't already set something.

There is one rough edge in the Global Styles border panel. Assuming theme.json does not set a border style: 
- Select a border width or color
- The border style gets set to `solid`
- Clear the width or color selected
- The border style remains set to `solid`

This could be polished in a follow-up as it likely requires some convoluted logic tracking user set vs theme.json values so that the style doesn't get cleared when there is a default width or color etc.

> Jay's feedback is still valid, though: if you're adding a border, you'll most likely also want to customize the color at least. That's what makes me feel like this is a single control, not multiple separate ones. 

I think we should push re-imagining the border controls to a separate PR. I agree they need that attention but I would like to see that not block this switch to the ToolsPanel if it can be avoided/separated.

ciampo commented:
I haven't followed this conversation (and all of the experimentation that happened in other draft PRs), but as I'm reading it, I though about another approach:

Why couldn't we assume that the default value for border-related properties are:
- `border-width`: `0`
- `border-style`: `solid`
- `border-color`: `currentColor` (this default value could also be omitted, since the border color matches the color already by default)

These values would serve as the default, initial value, but also as the values that are applied when resetting via `ToolsPanel`.

When assuming these defaults:

- the only control that the user would need to change to see a border appear becomes `border-width` — and therefore that control could be place in a `ToolsPanelItem` with `isShownByDefault={true}`
- the `border-style` and `border-color` controls would be used to tweak the border once it has appeared, and therefore could be optional


Apologies in advance if this approach has already been discussed and rejected

aaronrobertshaw commented:
> Why couldn't we assume that the default value for border-related properties are:
> 
> - border-width: 0
> - border-style: solid
> - border-color: currentColor (this default value could also be omitted, since the border color matches the color already by default)
> 

The issue here is we wouldn't know exactly what styles might have been set already via Global Styles. So if those were applied to the controls as defaults, they would override user selections which we should honor. @stacimc touches on this problem here: https://github.com/WordPress/gutenberg/pull/33743#issuecomment-974532011

This PR effectively applies those defaults you suggested but only via a `:where` CSS style so that it isn't too heavy handed in overriding global styles. My comments (https://github.com/WordPress/gutenberg/pull/33743#issuecomment-977593105) and (https://github.com/WordPress/gutenberg/pull/33743#issuecomment-977615151) hopefully help explain things.

ciampo commented:
> The issue here is we wouldn't know exactly what styles might have been set already via Global Styles. So if those were applied to the controls as defaults, they would override user selections which we should honor. @stacimc touches on this problem here: [#33743 (comment)](https://github.com/WordPress/gutenberg/pull/33743#issuecomment-974532011)
> 
> This PR effectively applies those defaults you suggested but only via a `:where` CSS style so that it isn't too heavy handed in overriding global styles. My comments ([#33743 (comment)](https://github.com/WordPress/gutenberg/pull/33743#issuecomment-977593105)) and ([#33743 (comment)](https://github.com/WordPress/gutenberg/pull/33743#issuecomment-977615151)) hopefully help explain things.

I had the suspicion that I was going to propose something already discussed, but I thought I'd give it a try anyway 😅 

Your explanation makes total sense, thank you!

stacimc commented:
The latest changes for setting default fallbacks tests great for me 🥳  I tested in the block editor and in the site editor, with and without settings coming from theme.json/global styles.

Something nice I noted: this differs from the approach from #34131 in that setting a color even in global styles will force the border to be visible for all blocks of that type. For example if you provide a blue border-color for the Group block in your `theme.json`, all Group blocks will immediately display blue solid borders unless you explicitly set their border-width to 0px.

I prefer it this way! You can still explicitly set `0px` width in your `theme.json` as well if your intention is only to set a default color, but not to opt all blocks into a visible border. And this makes the behavior completely consistent 👍

The panel is testing really well for me now. I'd love to get this in and iterate on the controls separately.

ramonjd commented:
Just retested this PR.

I think the addition of the `:where` selectors to force a solid style works intuitively and is a non-instrusive compromise for now while we iterate and test.

Things work as expected for me in the post and site editors. I can see the effects of modifying borders settings immediately, and can apply width and color to my borders without the style option engaged. 

When I don't specify a style in the post editor, setting the border style in global styles takes precedence.

I think it's good to go.

aaronrobertshaw commented:
Thank you everyone for testing the latest iteration of this! I appreciate it 🙇 

> The panel is testing really well for me now. I'd love to get this in and iterate on the controls separately.

> I think it's good to go.

Sounds like we're nearly ready to merge. 

@jasmussen or @jameskoster do you have any objections to us merging this PR and iterating on the border controls separately?

@ramonjd already has a PR up to discuss and address which border controls will show by default for individual blocks. https://github.com/WordPress/gutenberg/pull/34061

One final thing worth mentioning, is I think the [PR](https://github.com/WordPress/gutenberg/pull/36753) @stacimc created, splitting out the multiple origin color palette update to the border color control, should be merged before this as it might make it easier to sneak that into 5.9 as a fix.

jasmussen commented:
To take stock. This is trunk:
![current](https://user-images.githubusercontent.com/1204802/143413947-6b393058-24ca-4941-886c-24bd910e95c1.gif)

Border being something that should not be shown by default on any block (possibly with the exception of quotes in the future, if we manage to absorb the left border into theme.json), to me it then feels appropriate that this panel starts collapsed, but then also that it shows both width, style, and color at once when expanded. The big downside here is that you have to set both a width and a style for any border to show.

This branch:
![this branch](https://user-images.githubusercontent.com/1204802/143414518-0f39a02a-4836-47c5-8adb-732e5c905a09.gif)

Border being something you explicitly add using the "Plus" button feels good, and the fact that you can set a border width without also selecting a style is a key problem solved.

The fundamental problem here is that width and style are _two separate controls_ that both need to be added if you want to make a dashed line. While changing to dashed or dots isn't something I expect people to do that often, it still feels like those controls are separate _only_ because `border-style` and `border-width` are separate in CSS, as opposed to being good for the flow. This reminds me of the `font-style`and `font-family` support we added as a single dropdown that included both in one list.

In that vein, I still think that at least border-width and style (unsure about color) should be considered a single user interface control to add, rather than two separate ones. 

Can that happen in a followup? I think that depends mostly on whether we can come back to the panel and make it a single control in the future, without having to handle a slew of deprecations and related issues. If we can, seems fine to land.

One thing I noticed though — which is possibly/probably unrelated to this PR, is that the panel when empty has the wrong height:

<img width="333" alt="Screenshot 2021-11-25 at 10 18 51" src="https://user-images.githubusercontent.com/1204802/143416042-0eeda61f-0f02-4018-aef4-6239bd1aeb35.png">

It's 53px tall when it should be 48:
<img width="347" alt="Screenshot 2021-11-25 at 10 33 02" src="https://user-images.githubusercontent.com/1204802/143416123-0a7d3daa-c4da-4931-bcb7-68af9635fbff.png">

That height appears to be due to the component inheriting an arbitrary line-height from an ancestor. If we add `line-height: 0;` to the panel, it becomes 49px, the extra pixel probably explained by the border:

<img width="377" alt="Screenshot 2021-11-25 at 10 35 14" src="https://user-images.githubusercontent.com/1204802/143416507-d16c1543-7975-443f-9d96-f7ecece34151.png">

aaronrobertshaw commented:
Thanks for the detailed comment @jasmussen 👍 

> Border being something you explicitly add using the "Plus" button feels good, and the fact that you can set a border width without also selecting a style is a key problem solved.

It's not just border width you can set and immediately see a border. With this PR, you see a border regardless of which individual control you set, width, style or color.

> The fundamental problem here is that width and style are *two separate controls* that both need to be added if you want to make a dashed line

In my testing across a few different themes, this isn't the case. 

You can select only border style from the border panel menu, select a style and see a border without the need to set a width. The current approach in this PR should pick up browser defaults.

<img src="https://user-images.githubusercontent.com/60436221/143527866-681b6074-3474-4db9-811f-4f063b46f1b3.gif" width="500" />

> In that vein, I still think that at least border-width and style (unsure about color) should be considered a single user interface control to add, rather than two separate ones.

I have no objections to this being a follow-up. I do not think it should block this PR given you can create a border via any of the individual controls and this is a improvement over the current state.

If a mockup can be created for how a single control should look and function, I'll be happy to pick that up for implementation.

> Can that happen in a followup? I think that depends mostly on whether we can come back to the panel and make it a single control in the future, without having to handle a slew of deprecations and related issues. If we can, seems fine to land.

The border block supports (other than named colors) just add inline styles. Those inline styles are separate now and would continue to be so once combining controls together in the UI. There should be no deprecation issues as a result.

This is the same as the `font-style` and `font-weight` block supports that we combined under the "Appearance" control. The `font-family` support has its own individual UI.

The inline styles here would benefit from staying separate as per `font-style` and `font-weight` so blocks still have the freedom to opt-out of any individual border feature they did not want e.g. `border-style`.

> One thing I noticed though — which is possibly/probably unrelated to this PR, is that the panel when empty has the wrong height:

You are correct this is unrelated to this PR. When the `ToolsPanel` previously had a horizontal ellipsis for its menu icon, it didn't inflate the height when empty. When the menu button has its `line-height` set to `0` as suggested, this brings the empty state for the panel to `49px`. This matches the collapsed state of the other panels due to the border. The `ToolsPanel` is laid out differently so a direct comparison between the actual panel headers' heights can't be made.

I'll fix this height issue in a separate PR.

aaronrobertshaw commented:
I added the tweak to fix the empty `ToolsPanel` height in https://github.com/WordPress/gutenberg/pull/36895.

The plan now is to merge this Border panel PR first thing Monday.

jasmussen commented:
> In my testing across a few different themes, this isn't the case.
> You can select only border style from the border panel menu, select a style and see a border without the need to set a width. The current approach in this PR should pick up browser defaults.

Just in case there was any doubt, the fact that style and width work as separate controls is commendable. But what I meant was that you should never have to add both style and width with two visits to the tools panel. There should be only a single "Border" control. Something like this:

<img width="330" alt="tools panel" src="https://user-images.githubusercontent.com/1204802/143548225-c4264db8-9ead-4ab2-9d59-b5145bc99895.png">

> If a mockup can be created for how a single control should look and function, I'll be happy to pick that up for implementation.

I think there are still some border designs underway for improving how that looks in the future. So in the mean time and just for simplicity, we can use the current design:

<img width="291" alt="Screenshot 2021-11-26 at 09 18 35" src="https://user-images.githubusercontent.com/1204802/143549359-0727451e-f811-4ad3-83a3-7fb272649f19.png">

☝️ That could be a single control called "Border" in the tools panel.

You should never see only the width, or only the style. Does that sound okay?

> The inline styles here would benefit from staying separate as per font-style and font-weight so blocks still have the freedom to opt-out of any individual border feature they did not want e.g. border-style.

Sounds good. The key is that it doesn't make a lot of sense to add only width and not also style, or only style and not also width.

aaronrobertshaw commented:
> I think there are still some border designs underway for improving how that looks in the future. So in the mean time and just for simplicity, we can use the current design:
> 
> ☝️ That could be a single control called "Border" in the tools panel.
> 
> You should never see only the width, or only the style. Does that sound okay?


> Sounds good. The key is that it doesn't make a lot of sense to add only width and not also style, or only style and not also width.

I could see a theme author wishing to disable border style to prevent dotted/dashed borders etc but still allowing variations in border width.

I'll continue with the plan to merge this PR today and work on a separate PR to combine the width/style controls within the `ToolsPanel` menu only. We can discuss that further there.

youknowriad commented on packages/edit-site/src/components/global-styles/border-panel.js:
```diff
@@ -77,70 +90,121 @@ export default function BorderPanel( { name } ) {
+	// can see a visible border.
+	const handleOnChangeWithStyle = ( setStyle ) => ( value ) => {
+		if ( !! value && ! borderStyle ) {
+			setBorderStyle( 'solid' );
```
Do you happen to know why we do that here? This is creating some issues.

aaronrobertshaw replied:
There was feedback here that we needed to ensure when first setting a border width or color that there is a visible border. That involved ensuring a border style was set.

The inline comment above this line and the commit message point to that as the primary reason (https://github.com/WordPress/gutenberg/pull/33743/commits/3a8e7483ec9d1e2cf2e6822f1ac2a1077c035a24).
